File: packages/js-sdk/src/sandbox/process.ts
import { PlainMessage } from '@bufbuild/protobuf'
import {
  createPromiseClient,
  PromiseClient,
  Transport,
} from '@connectrpc/connect'

import { ProcessService } from '../envd/process/v1/process_connect'
import {
  ProcessConfig as PsProcessConfig,
  ProcessEvent_EndEvent,
  ConnectResponse,
  ConnectRequest,
  SendInputRequest,
  StreamInputRequest,
  Signal,
  StartRequest,
  StartResponse,
  UpdateRequest,
} from '../envd/process/v1/process_pb'
import { ConnectionOpts } from '../connectionConfig'

export interface ProcessOutput {
  stdout?: string
  stderr?: string
}

export class ProcessHandle {
  private readonly end: Promise<{
    exit: ProcessEvent_EndEvent,
    stdout: Uint8Array[],
    stderr: Uint8Array[],
  }>

  constructor(
    readonly pid: number,
    readonly handleDisconnect: () => void,
    private readonly handleKill: () => Promise<void>,
    private readonly events: AsyncIterable<ConnectResponse | StartResponse>,
    private readonly handleStdout?: ((data: string) => void | Promise<void>),
    private readonly handleStderr?: ((data: string) => void | Promise<void>),
    private readonly handlePtyData?: ((data: Uint8Array) => void | Promise<void>),
  ) {
    this.end = this.handleEvents()
  }

  private static joinUint8Arrays(arrays: Uint8Array[]) {
    const length = arrays.reduce((acc, cur) => acc + cur.length, 0)
    const result = new Uint8Array(length)
    let offset = 0

    for (const array of arrays) {
      result.set(array, offset)
      offset += array.length
    }

    return result
  }

  disconnect() {
    this.handleDisconnect()
  }

  async wait(): Promise<ProcessResult> {
    const result = await this.end

    return {
      exitCode: result.exit.exitCode,
      error: result.exit.error,
      stdout: ProcessHandle.joinUint8Arrays(result.stdout).toString(),
      stderr: ProcessHandle.joinUint8Arrays(result.stderr).toString(),
    }
  }

  async kill() {
    this.disconnect()
    await this.handleKill()
  }

  private async handleEvents() {
    const stdout: Uint8Array[] = []
    const stderr: Uint8Array[] = []

    for await (const event of this.events) {
      switch (event.event?.event.case) {
        case 'data':
          switch (event.event.event.value.output.case) {
            case 'stdout':
              stdout.push(event.event.event.value.output.value)
              await this.handleStdout?.(event.event.event.value.output.value.toString())
              break
            case 'stderr':
              stderr.push(event.event.event.value.output.value)
              await this.handleStderr?.(event.event.event.value.output.value.toString())
              break
            case 'pty':
              await this.handlePtyData?.(event.event.event.value.output.value)
              break
          }
          break
        case 'end':
          return {
            exit: event.event.event.value,
            stdout,
            stderr,
          }
      }
    }

    throw new Error('Process ended without an end event')
  }
}

export interface ProcessResult extends ProcessOutput {
  exitCode: number
  error?: string
}

export type ProcessConfig = PlainMessage<PsProcessConfig>

// TODO: Sane defaults
// TODO: Add timeout to wait
// TODO: How to handle start + wait
// TODO: Move cwd resolution to envd?
// TODO: Should cwd be calculated in envd?
// TODO: More optional fields?
// TODO: Add req timeout
// TODO: Enable using process handle as iterable
// TODO: Solve stream input
// TODO: For watch and other stream ensure the timeout throw is handlable
// TODO: Create class from process handler
// TODO: Disconnect during kill
// TODO: Add iterator returns (python too)
// TODO: Passing timeout to envd context via metadata?
// TODO: Finish requestTimeoutMs handling in normal API
// TODO: Does the abort controller work when we specify the request timeout via connect rpc?
export class Process {
  protected readonly service: PromiseClient<typeof ProcessService> = createPromiseClient(ProcessService, this.transport)

  constructor(private readonly transport: Transport) { }

  async list(opts?: Pick<ConnectionOpts, 'requestTimeoutMs'>): Promise<ProcessConfig[]> {
    const res = await this.service.list({}, {
      timeoutMs: opts?.requestTimeoutMs,
    })
    return res.processes
  }

  async start(
    cmd: string,
    opts: {
      cwd?: string,
      user?: 'root' | 'user',
      envs?: Record<string, string>,
      onStdout?: ((data: string) => void | Promise<void>),
      onStderr?: ((data: string) => void | Promise<void>),
    } & Pick<ConnectionOpts, 'requestTimeoutMs'>
  ): Promise<ProcessHandle> {
    const params: PlainMessage<StartRequest> = {
      owner: {
        credential: {
          case: 'username',
          value: opts.user || 'user',
        },
      },
      process: {
        cmd: '/bin/bash',
        cwd: opts.cwd,
        envs: opts.envs || {},
        args: ['-l', '-c', cmd],
      },
    }

    const controller = new AbortController()

    const events = this.service.start(params, {
      signal: controller.signal,
      timeoutMs: opts.requestTimeoutMs,
    })

    const startEvent: StartResponse = (await events[Symbol.asyncIterator]().next()).value

    if (startEvent.event?.event.case !== 'start') {
      throw new Error('Expected start event')
    }

    const pid = startEvent.event.event.value.pid

    return new ProcessHandle(
      pid,
      () => controller.abort(),
      () => this.kill(pid),
      events,
      opts.onStdout,
      opts.onStderr,
    )
  }

  async connect(
    pid: number,
    opts?: {
      onStdout?: ((data: string) => void | Promise<void>),
      onStderr?: ((data: string) => void | Promise<void>),
    } & Pick<ConnectionOpts, 'requestTimeoutMs'>
  ): Promise<ProcessHandle> {
    const params: PlainMessage<ConnectRequest> = {
      process: {
        selector: {
          case: 'pid',
          value: pid,
        }
      },
    }
    const controller = new AbortController()

    const events = this.service.connect(params, {
      signal: controller.signal,
      timeoutMs: opts?.requestTimeoutMs,
    })

    return new ProcessHandle(
      pid,
      () => controller.abort(),
      () => this.kill(pid),
      events,
      opts?.onStdout,
      opts?.onStderr,
    )
  }

  async kill(pid: number, opts?: Pick<ConnectionOpts, 'requestTimeoutMs'>): Promise<void> {
    await this.service.sendSignal({
      process: {
        selector: {
          case: 'pid',
          value: pid,
        }
      },
      signal: Signal.SIGKILL,
    }, {
      timeoutMs: opts?.requestTimeoutMs,
    })
  }

  async sendStdin(pid: number, data: Uint8Array, opts?: Pick<ConnectionOpts, 'requestTimeoutMs'>): Promise<void> {
    const params: PlainMessage<SendInputRequest> = {
      process: {
        selector: {
          case: 'pid',
          value: pid,
        }
      },
      input: {
        input: {
          case: 'stdin',
          value: data,
        },
      },
    }

    await this.service.sendInput(params, {
      timeoutMs: opts?.requestTimeoutMs,
    })
  }

  async startTerminal({ cols, rows, onData }: {
    cols: number,
    rows: number,
    onData: (data: Uint8Array) => void | Promise<void>,
  }, opts?: Pick<ConnectionOpts, 'requestTimeoutMs'>): Promise<ProcessHandle> {
    const params: PlainMessage<StartRequest> = {
      owner: {
        credential: {
          case: 'username',
          value: 'user',
        },
      },
      process: {
        cmd: '/bin/bash',
        envs: {},
        args: ['-i', '-l'],
      },
      pty: {
        size: {
          cols,
          rows,
        },
      },
    }

    const controller = new AbortController()

    const events = this.service.start(params, {
      timeoutMs: opts?.requestTimeoutMs,
      signal: controller.signal,
    })

    const startEvent: StartResponse = (await events[Symbol.asyncIterator]().next()).value

    if (startEvent.event?.event.case !== 'start') {
      throw new Error('Expected start event')
    }

    const pid = startEvent.event.event.value.pid

    return new ProcessHandle(
      pid,
      () => controller.abort(),
      () => this.kill(pid),
      events,
      undefined,
      undefined,
      onData,
    )
  }

  async resizeTerminal(
    pid: number,
    size: {
      cols: number,
      rows: number,
    },
    opts?: Pick<ConnectionOpts, 'requestTimeoutMs'>,
  ): Promise<void> {
    const params: PlainMessage<UpdateRequest> = {
      process: {
        selector: {
          case: 'pid',
          value: pid,
        }
      },
      pty: {
        size,
      },
    }

    await this.service.update(params, {
      timeoutMs: opts?.requestTimeoutMs,
    })
  }

  private async streamTerminalInput(params: AsyncIterable<PlainMessage<StreamInputRequest>>, opts?: Pick<ConnectionOpts, 'requestTimeoutMs'>) {
    const controller = new AbortController()

    this.service.streamInput(params, {
      signal: controller.signal,
      timeoutMs: opts?.requestTimeoutMs,
    })

    return {
      stop: () => controller.abort(),
    }
  }
}

File: packages/python-sdk/e2b/sandbox/process.py
import logging
import connect

from typing import Dict, Optional, Callable, Any, Generator, Union

from envd.process.v1 import process_connect, process_pb2

logger = logging.getLogger(__name__)


class ProcessOutput:
    def __init__(self, stdout: Optional[bytes], stderr: Optional[bytes]) -> None:
        self.stdout = str(stdout) if stdout else None
        self.stderr = str(stderr) if stderr else None


class ProcessHandle:
    def __init__(
        self,
        pid: int,
        kill: Callable[[], None],
        events: Generator[
            Union[process_pb2.StartResponse, process_pb2.ConnectResponse], Any, None
        ],
    ):
        self.pid = pid
        self.kill = kill
        self._events = events

        self._stdout: bytes = b""
        self._stderr: bytes = b""

        self._end_event: Union[process_pb2.ProcessEvent.EndEvent, None] = None

    def __next__(self):
        event = next(self._events)

        if event.HasField("data"):
            if event.event.data.stdout:
                self._stdout += event.event.data.stdout
                return ProcessOutput(self._stdout, None)
            if event.event.data.stderr:
                self._stderr += event.event.data.stderr
                return ProcessOutput(None, self._stderr)
        if event.HasField("end"):
            self._end_event = event.event.end

        raise StopIteration

    def __iter__(self):
        return self

    def wait(self):
        for _ in self:
            pass

        if self._end_event is None:
            raise RuntimeError("Process has not ended")

        return ProcessResult(
            stdout=self._stdout,
            stderr=self._stderr,
            exit_code=self._end_event.exit_code,
            error=self._end_event.error,
        )


class ProcessResult(ProcessOutput):
    def __init__(
        self,
        stdout: bytes,
        stderr: bytes,
        exit_code: int,
        error: Optional[str] = None,
    ) -> None:
        super().__init__(stdout, stderr)
        self.exit_code = exit_code
        self.error = error


# TODO: Add disconnect for process handle
class Process:
    def __init__(self, base_url: str) -> None:
        self.base_url = base_url

        self._service = process_connect.ProcessServiceClient(
            self.base_url,
            compressor=connect.GzipCompressor,
        )

    def list(self):
        params = process_pb2.ListRequest()

        res = self._service.list(params)
        return [p for p in res.processes]

    def kill(self, pid: int):
        params = process_pb2.SendSignalRequest(
            process=process_pb2.ProcessSelector(pid=pid),
            signal=process_pb2.Signal.SIGNAL_SIGKILL,
        )

        self._service.send_signal(params)

    def sendStdin(self, pid: int, data: bytes):
        params = process_pb2.SendInputRequest(
            process=process_pb2.ProcessSelector(pid=pid),
            input=process_pb2.ProcessInput(stdin=data),
        )

        self._service.send_input(params)

    def start(
        self,
        cmd: str,
        envs: Optional[Dict[str, str]] = {},
        user: str = "user",
        cwd: Optional[str] = None,
    ):
        params = process_pb2.StartRequest(
            owner=process_pb2.Credential(username=user),
            process=process_pb2.ProcessConfig(
                cmd=cmd,
                envs=envs,
                cwd=cwd,
            ),
        )

        events = self._service.start(params)

        start_event = next(events)

        return ProcessHandle(
            pid=start_event.event.start.pid,
            kill=lambda: self.kill(start_event.event.start.pid),
            events=events,
        )

    def connect(self, pid: int):
        params = process_pb2.ConnectRequest(
            process=process_pb2.ProcessSelector(pid=pid),
        )

        events = self._service.connect(params)

        return ProcessHandle(
            pid=pid,
            kill=lambda: self.kill(pid),
            events=events,
        )
